File: mobile/modules/acs-meeting/android/src/main/java/com/mentra/acsmeeting/audio/AudioPolicyApplier.kt
package com.mentra.acsmeeting.audio

interface AudioStreamController {
  fun readActive(): ActiveStreamKind
  fun isPhysicallyMuted(): Boolean?
  fun setGlassesPcmEnabled(enabled: Boolean)
  fun setPhonePcmEnabled(enabled: Boolean)
  fun mutePhysical(): Result<Unit>
  fun unmutePhysical(): Result<Unit>
  fun stopActive(): Result<Unit>
}

interface PolicyScheduler {
  fun schedule(delayMs: Long, task: () -> Unit)
  fun cancelPending()
}

class AudioPolicyApplier(
  private val controller: AudioStreamController,
  private val scheduler: PolicyScheduler,
  private val log: (String) -> Unit = {},
) {
  private var generation = 0L
  private var lastDecision: AudioPolicyDecision? = null
  private var lastSafety: AudioSafety = AudioSafety.DEGRADED

  fun lastSafety(): AudioSafety = lastSafety

  fun apply(desired: AudioSourceKind, userMuted: Boolean, reason: String): AudioSafety {
    generation += 1
    scheduler.cancelPending()
    return runApply(desired, userMuted, reason, generation, force = false, armConvergence = true)
  }

  fun reset() {
    generation += 1
    scheduler.cancelPending()
    lastDecision = null
    lastSafety = AudioSafety.DEGRADED
  }

  private fun runApply(
    desired: AudioSourceKind,
    userMuted: Boolean,
    reason: String,
    gen: Long,
    force: Boolean,
    armConvergence: Boolean,
  ): AudioSafety {
    if (gen != generation) return lastSafety
    val active = controller.readActive()
    val decision = AcsAudioPolicy.decide(desired, active, userMuted)
    val outcome = applyDecision(decision, force)
    if (gen != generation) return lastSafety

    val after = controller.readActive()
    val expected = AcsAudioPolicy.expectedStream(desired)
    val safety = safetyOf(desired, after, outcome)
    lastDecision = decision
    lastSafety = safety
    log(
      "audio policy reason=$reason source=$desired userMuted=$userMuted " +
        "active=$after expected=$expected glassesPcm=${decision.glassesPcmEnabled} " +
        "phonePcm=${decision.phonePcmEnabled} mute=${decision.physicalMute} safety=$safety",
    )

    if (after == expected) {
      scheduler.cancelPending()
    } else if (armConvergence && gen == generation) {
      armConvergence(desired, userMuted, gen, expected)
    }
    return lastSafety
  }

  private fun armConvergence(
    desired: AudioSourceKind,
    userMuted: Boolean,
    gen: Long,
    expected: ActiveStreamKind,
  ) {
    CONVERGE_DELAYS_MS.forEachIndexed { index, delayMs ->
      scheduler.schedule(delayMs) {
        if (gen != generation) return@schedule
        val isLast = index == CONVERGE_DELAYS_MS.lastIndex
        runApply(desired, userMuted, "converge-${delayMs}ms", gen, force = true, armConvergence = false)
        if (isLast && gen == generation && controller.readActive() != expected) {
          log("audio policy convergence timed out source=$desired expected=$expected holding silence")
          lastSafety = AudioSafety.DEGRADED
        }
      }
    }
  }

  private fun applyDecision(decision: AudioPolicyDecision, force: Boolean): EffectOutcome {
    val previous = lastDecision
    val skipGlassesPcm = !force && previous != null && previous.glassesPcmEnabled == decision.glassesPcmEnabled
    val skipPhonePcm = !force && previous != null && previous.phonePcmEnabled == decision.phonePcmEnabled
    val skipMute = !force && previous != null && previous.physicalMute == decision.physicalMute
    if (!skipGlassesPcm) {
      controller.setGlassesPcmEnabled(decision.glassesPcmEnabled)
    }
    if (!skipPhonePcm) {
      controller.setPhonePcmEnabled(decision.phonePcmEnabled)
    }
    if (skipMute) return EffectOutcome.NONE

    return when (decision.physicalMute) {
      PhysicalMuteAction.LEAVE_ALONE -> EffectOutcome.NONE
      PhysicalMuteAction.UNMUTE -> {
        val result = controller.unmutePhysical()
        val unmuted = controller.isPhysicallyMuted()
        val failed = result.isFailure || unmuted == true
        EffectOutcome(unmuteAttempted = true, unmuteFailed = failed)
      }
      PhysicalMuteAction.MUTE -> {
        val result = controller.mutePhysical()
        val muted = controller.isPhysicallyMuted()
        val muteFailed = result.isFailure || muted == false
        if (!muteFailed) return EffectOutcome(muteAttempted = true)
        val stop = controller.stopActive()
        val stopped = controller.readActive() == ActiveStreamKind.NONE
        EffectOutcome(
          muteAttempted = true,
          muteFailed = true,
          stopAttempted = true,
          stopFailed = stop.isFailure || !stopped,
        )
      }
    }
  }

  private fun safetyOf(
    desired: AudioSourceKind,
    active: ActiveStreamKind,
    outcome: EffectOutcome,
  ): AudioSafety {
    if (outcome.stopFailed) return AudioSafety.UNSAFE
    if (outcome.stopAttempted) return AudioSafety.SAFE
    if (outcome.unmuteFailed) return AudioSafety.DEGRADED
    val expected = AcsAudioPolicy.expectedStream(desired)
    if (active != expected) return AudioSafety.DEGRADED
    return AudioSafety.SAFE
  }

  private data class EffectOutcome(
    val muteAttempted: Boolean = false,
    val muteFailed: Boolean = false,
    val unmuteAttempted: Boolean = false,
    val unmuteFailed: Boolean = false,
    val stopAttempted: Boolean = false,
    val stopFailed: Boolean = false,
  ) {
    companion object {
      val NONE = EffectOutcome()
    }
  }

  companion object {
    val CONVERGE_DELAYS_MS = longArrayOf(50, 100, 200, 400, 800, 1600)
  }
}

class ExecutorPolicyScheduler(
  private val executor: java.util.concurrent.ScheduledExecutorService,
) : PolicyScheduler {
  private val pending = java.util.concurrent.CopyOnWriteArrayList<java.util.concurrent.ScheduledFuture<*>>()

  override fun schedule(delayMs: Long, task: () -> Unit) {
    pending.add(executor.schedule(task, delayMs, java.util.concurrent.TimeUnit.MILLISECONDS))
  }

  override fun cancelPending() {
    pending.forEach { it.cancel(false) }
    pending.clear()
  }
}

File: mobile/modules/acs-meeting/ios/PolicyKit/Sources/AcsAudioPolicy/AudioPolicyApplier.swift
public protocol AudioStreamController {
  func readActive() -> ActiveStreamKind
  func isPhysicallyMuted() -> Bool?
  func setGlassesPcmEnabled(_ enabled: Bool)
  func setPhonePcmEnabled(_ enabled: Bool)
  func mutePhysical() -> Result<Void, Error>
  func unmutePhysical() -> Result<Void, Error>
  func stopActive() -> Result<Void, Error>
}

public protocol PolicyScheduler: AnyObject {
  func schedule(delayMs: Int, task: @escaping () -> Void)
  func cancelPending()
}

public final class AudioPolicyApplier {
  public static let convergeDelaysMs = [50, 100, 200, 400, 800, 1600]

  private let controller: AudioStreamController
  private let scheduler: PolicyScheduler
  private let log: (String) -> Void
  private var generation: UInt64 = 0
  private var lastDecision: AudioPolicyDecision?
  private var lastSafetyValue: AudioSafety = .degraded

  public init(controller: AudioStreamController, scheduler: PolicyScheduler, log: @escaping (String) -> Void = { _ in }) {
    self.controller = controller
    self.scheduler = scheduler
    self.log = log
  }

  public func lastSafety() -> AudioSafety { lastSafetyValue }

  @discardableResult
  public func apply(desired: AudioSourceKind, userMuted: Bool, reason: String) -> AudioSafety {
    generation += 1
    scheduler.cancelPending()
    return runApply(desired: desired, userMuted: userMuted, reason: reason, gen: generation, force: false, armConvergence: true)
  }

  public func reset() {
    generation += 1
    scheduler.cancelPending()
    lastDecision = nil
    lastSafetyValue = .degraded
  }

  @discardableResult
  private func runApply(
    desired: AudioSourceKind,
    userMuted: Bool,
    reason: String,
    gen: UInt64,
    force: Bool,
    armConvergence: Bool
  ) -> AudioSafety {
    guard gen == generation else { return lastSafetyValue }
    let active = controller.readActive()
    let decision = AcsAudioPolicy.decide(desired: desired, active: active, userMuted: userMuted)
    let outcome = applyDecision(decision, force: force)
    guard gen == generation else { return lastSafetyValue }

    let after = controller.readActive()
    let expected = AcsAudioPolicy.expectedStream(desired)
    let safety = safetyOf(desired: desired, active: after, outcome: outcome)
    lastDecision = decision
    lastSafetyValue = safety
    log(
      "audio policy reason=\(reason) source=\(desired) userMuted=\(userMuted) active=\(after) expected=\(expected) glassesPcm=\(decision.glassesPcmEnabled) phonePcm=\(decision.phonePcmEnabled) mute=\(decision.physicalMute) safety=\(safety)"
    )

    if after == expected {
      scheduler.cancelPending()
    } else if armConvergence && gen == generation {
      self.armConvergence(desired: desired, userMuted: userMuted, gen: gen, expected: expected)
    }
    return lastSafetyValue
  }

  private func armConvergence(desired: AudioSourceKind, userMuted: Bool, gen: UInt64, expected: ActiveStreamKind) {
    for (index, delayMs) in Self.convergeDelaysMs.enumerated() {
      scheduler.schedule(delayMs: delayMs) { [weak self] in
        guard let self, gen == self.generation else { return }
        let isLast = index == Self.convergeDelaysMs.count - 1
        _ = self.runApply(
          desired: desired,
          userMuted: userMuted,
          reason: "converge-\(delayMs)ms",
          gen: gen,
          force: true,
          armConvergence: false
        )
        if isLast && gen == self.generation && self.controller.readActive() != expected {
          self.log("audio policy convergence timed out source=\(desired) expected=\(expected) holding silence")
          self.lastSafetyValue = .degraded
        }
      }
    }
  }

  private func applyDecision(_ decision: AudioPolicyDecision, force: Bool) -> EffectOutcome {
    let previous = lastDecision
    let skipGlassesPcm = !force && previous?.glassesPcmEnabled == decision.glassesPcmEnabled
    let skipPhonePcm = !force && previous?.phonePcmEnabled == decision.phonePcmEnabled
    let skipMute = !force && previous?.physicalMute == decision.physicalMute
    if !skipGlassesPcm {
      controller.setGlassesPcmEnabled(decision.glassesPcmEnabled)
    }
    if !skipPhonePcm {
      controller.setPhonePcmEnabled(decision.phonePcmEnabled)
    }
    if skipMute { return .none }

    switch decision.physicalMute {
    case .leaveAlone:
      return .none
    case .unmute:
      let result = controller.unmutePhysical()
      let unmuted = controller.isPhysicallyMuted()
      let failed = result.isFailure || unmuted == true
      return EffectOutcome(unmuteAttempted: true, unmuteFailed: failed)
    case .mute:
      let result = controller.mutePhysical()
      let muted = controller.isPhysicallyMuted()
      let muteFailed = result.isFailure || muted == false
      if !muteFailed { return EffectOutcome(muteAttempted: true) }
      let stop = controller.stopActive()
      let stopped = controller.readActive() == .none
      return EffectOutcome(
        muteAttempted: true,
        muteFailed: true,
        stopAttempted: true,
        stopFailed: stop.isFailure || !stopped
      )
    }
  }

  private func safetyOf(desired: AudioSourceKind, active: ActiveStreamKind, outcome: EffectOutcome) -> AudioSafety {
    if outcome.stopFailed { return .unsafe }
    if outcome.stopAttempted { return .safe }
    if outcome.unmuteFailed { return .degraded }
    if active != AcsAudioPolicy.expectedStream(desired) { return .degraded }
    return .safe
  }
}

private struct EffectOutcome {
  var muteAttempted = false
  var muteFailed = false
  var unmuteAttempted = false
  var unmuteFailed = false
  var stopAttempted = false
  var stopFailed = false
  static let none = EffectOutcome()
}

private extension Result {
  var isFailure: Bool {
    if case .failure = self { return true }
    return false
  }
}
